src/parser/parse_type.hpp

```diff
@@ -0,0 +1,89 @@
+#ifndef _PARSER_TYPE_HPP_
+#define _PARSER_TYPE_HPP_
+
+#include "parse_tree.h"
+
+/* generic types */
+typedef enum pt_generic_type_enum
+{
+  PT_GENERIC_TYPE_NONE,
+  PT_GENERIC_TYPE_STRING,	/* any type of string */
+  PT_GENERIC_TYPE_STRING_VARYING,	/* VARCHAR or VARNCHAR */
+  PT_GENERIC_TYPE_CHAR,		/* VARCHAR or CHAR */
+  PT_GENERIC_TYPE_NCHAR,	/* VARNCHAR or NCHAR */
+  PT_GENERIC_TYPE_BIT,		/* BIT OR VARBIT */
+  PT_GENERIC_TYPE_DISCRETE_NUMBER,	/* SMALLINT, INTEGER, BIGINTEGER */
+  PT_GENERIC_TYPE_NUMBER,	/* any number type */
+  PT_GENERIC_TYPE_DATE,		/* date, datetime or timestamp */
+  PT_GENERIC_TYPE_DATETIME,	/* any date or time type */
+  PT_GENERIC_TYPE_SEQUENCE,	/* any type of sequence */
+  PT_GENERIC_TYPE_LOB,		/* BLOB or CLOB */
+  PT_GENERIC_TYPE_QUERY,	/* Sub query (for range operators) */
+  PT_GENERIC_TYPE_PRIMITIVE,	/* primitive types */
+  PT_GENERIC_TYPE_ANY,		/* any type */
+  PT_GENERIC_TYPE_JSON_VAL,     //PT_TYPE_VARCHAR, PT_TYPE_VARCHAR, PT_TYPE_CHAR, PT_TYPE_NCHAR, PT_TYPE_NUMERIC, PT_TYPE_INTEGER, PT_TYPE_LOGICAL, PT_TYPE_DOUBLE, PT_TYPE_NUMERIC, PT_TYPE_JSON, PT_TYPE_MAYBE, PT_TYPE_NULL
+  PT_GENERIC_TYPE_JSON_DOC,     //PT_TYPE_VARCHAR, PT_TYPE_VARNCHAR, PT_TYPE_CHAR, PT_TYPE_NCHAR, PT_TYPE_JSON, PT_TYPE_MAYBE, PT_TYPE_NULL
+  PT_GENERIC_TYPE_JSON_PATH,    //PT_TYPE_VARCHAR, PT_TYPE_VARNCHAR, PT_TYPE_CHAR, PT_TYPE_NCHAR, PT_TYPE_BIT, PT_TYPE_VARBIT, PT_TYPE_MAYBE, PT_TYPE_NULL
+} PT_GENERIC_TYPE_ENUM;
+
+
+/* expression argument type */
+struct pt_arg_type
+{
+  enum {NORMAL, GENERIC, INDEX} type;
```

Comment:
using 'type' here as data member name may be confusing.
I don't have any acceptable suggestion; I can think of "arg_class", "mode", "arg_origin".

Reply:
I called it "selector" initially, but Remzi preferred "type"
I would also eliminate pt_arg_type.val but this will increase the number of changes in this PR making the review more difficult; I'll let it for a following PR